### PR TITLE
feat: google drive permissions sync

### LINKED ISSUE #284: Google drive permissions management
we should sync emails from teams onto access to folders

### PR DESCRIPTION
**Pull request by bobheadxi**
# Pull Request

## Description

**Give a brief description of your changes:**

Enables an optional step to `team refresh`, enabled by providing `GCP_SERVICE_ACCOUNT_CREDENTIALS`, to synchronize team folders by sharing each team's configured `folder` to all emails provided by that team's members.

The current _intended_ use case for this is _only_ for the top-level folders indicated in https://docs.ubclaunchpad.com/handbook/tools/drive - i.e I will be assigning the following folders:

* `all`: `Projects`
* `leads`: `UBC Launch Pad`
* `design`: `Design`
* `strategy`: `Strategy`

As such, no care has been taken for certain edge cases (ie "what if I *don't* want to unshare with anyone not on a rocket team?") - for this iteration of this feature, I am going to keep it simple.

The goal is to remove one of the last steps that leads have to do from onboarding - see https://github.com/ubclaunchpad/docs/pull/175 . It currently requires a refresh - see https://github.com/ubclaunchpad/rocket2/issues/490, https://github.com/ubclaunchpad/rocket2/issues/495

## Testing

**If testing this change requires extra setup, please document it here:**

Simple unit test provided. ~Working on a clicky-click test as well.~

Update - verified this works!

image

## Ticket(s)

Closes #284 

(Create a copy of that line for each Github Issue affected,
and replace "Affects" with "Closes" if merging this will close the relevant ticket.)

### CODE REVIEW

File: app/controller/command/commands/team.py [hunk 1]
```diff
@@ -650,3 +669,30 @@ def refresh_all_team(self):
...
+            for m in t.members:
+                if len(m.email) > 0:
+                    emails.append(m.email)
+
+            if len(emails) > 0:
+                logging.info("Synchronizing permissions for "
+                             + f"{t.github_team_name}'s folder ({t.folder}) "
+                             + f"to {emails}")
+                self.gcp.set_drive_permissions(
+                    t.folder, t.github_team_name, emails)
```
Comment (cheukyin699) [inline app/controller/command/commands/team.py:698]: You may want to run a more specialized version of this function, specific to the team that called with a `--folder`, to sync permissions for that team's team members only.

Something similar to what happens when we do `/rocket user edit --github` and we do an invite for that github user to join our organization.

File: app/model/team.py [hunk 2]
```diff
@@ -31,6 +32,7 @@ def get_attachment(self):
             ('Github Team Name', self.github_team_name),
             ('Display Name', self.display_name),
             ('Platform', self.platform),
+            ('Folder', self.folder),
```
Comment (cheukyin699) [inline app/model/team.py:35]: You will also need to modify the `Team.to_dict()` function so that we can properly store the folder.

**Review by chuck-sys — CHANGES_REQUESTED** [review]
Needs a bit more testing.

File: app/model/team.py (hunk 2 above)
Comment (bobheadxi) [reply app/model/team.py:35]: c5d65981dd8967ead7789563feea61b5c2159f28

File: app/controller/command/commands/team.py (hunk 1 above)
Comment (bobheadxi) [reply app/controller/command/commands/team.py:698]: Do you think it's necessary at the moment? Happy to implement this in a follow-up instead :)

**Review by chuck-sys — COMMENTED** [review]
Okay so here's the thing. `/rocket team refresh` takes around 20 seconds to execute (last I checked). The more members we have in Launch Pad, the longer it takes, `O(n)`. We'll still need ways for users to receive the permission sharing automatically, whether if it is a user just adding their emails, or a team adding a folder ID into the system.

I am very adamant about **not** using `/rocket team refresh` unless we really need to because ... there are lots of side effects that I don't know about, that may be disruptive. For example, today I found out that it would forcefully add a user back into the organization if the user was in ddb but was already out of the org (one possibility would be if the user ID was still in a team in ddb).

Will have to implement this as a follow-up. For now I'm approving this.

**Review by chuck-sys — APPROVED** [review]
Okay so here's the thing. `/rocket team refresh` takes around 20 seconds to execute (last I checked). The more members we have in Launch Pad, the longer it takes, `O(n)`. We'll still need ways for users to receive the permission sharing automatically, whether if it is a user just adding their emails, or a team adding a folder ID into the system.

I am very adamant about **not** using `/rocket team refresh` unless we really need to because ... there are lots of side effects that I don't know about, that may be disruptive. For example, today I found out that it would forcefully add a user back into the organization if the user was in ddb but was already out of the org (one possibility would be if the user ID was still in a team in ddb).

Will have to implement this as a follow-up. For now I'm approving this.

**Comment by bobheadxi** [comment]
@cheukyin699 agreed and noted! I've created https://github.com/ubclaunchpad/rocket2/issues/490 to cover that

**Pull request merged**